File: leetcode/20.cc
class Solution {
public:
    bool isValid(string s) {
        
        stack<char> st;

        for (int i=0; i<s.size(); i++) {
            if (st.empty() || !IsMatch(st.top(), s[i])) {
                st.push(s[i]);
            } else {
                st.pop();
            }
        }
        
        return st.empty();
    }
private:
    bool IsMatch(char a, char b) {
        if (a == '(' && b == ')') return true;
        if (a == '[' && b == ']') return true;
        if (a == '{' && b == '}') return true;
        return false;
    }
};

File: leetcode/20.py
''' 20. Valid Parentheses '''
class Solution:
    def isValid(self, s: str) -> bool:
        pair:dict = {
            "[": "]",
            "{": "}",
            "(": ")"
        }
        dq = deque()
        
        for ch in s:
            if ch in pair:
                dq.append(ch)
            else:
                if len(dq) <= 0 or pair[dq[-1]] != ch: return False
                dq.pop()
        return len(dq) == 0
        
